R/mlPrediction.R

```diff
@@ -0,0 +1,261 @@
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+mlPrediction <- function(jaspResults, dataset, options, ...) {
+  
+  # Preparatory work
+  model 	<- .mlPredictionReadModel(options)
+  dataset 	<- .mlPredictionReadData(options)
+  type 		<- .mlPredictionModelType(model)
+  
+  # Check if analysis is ready
+  ready <- .mlPredictionReady(model, dataset, options)
+  
+  # Create a table summarizing the loaded model
+  .mlPredictionModelSummaryTable(model, dataset, options, jaspResults, ready, position = 1)
+  
+  # Create a table containing the predicted values and predictors
+  .mlPredictionsTable(model, dataset, options, type, jaspResults, ready, position = 2)
+  
+  # Add predicted outcomes to data set
+  .mlPredictionsAddPredictions(model, dataset, options, jaspResults, ready)
+  
+}
+
+.mlPredictionReadModel <- function(options){
+	  if(options[["file"]] != ""){
+    model <- try({ readRDS(options[["file"]]) })
+    if(!(class(model) %in% c("lda", "gbm", "randomForest", "cv.glmnet")))
+      jaspBase:::.quitAnalysis(gettextf("The imported model (type: %1$s) is currently not supported in JASP.", class(model)))
+  } else {
+    model <- NULL
+  }
+  return(model)
+}
+
+.mlPredictionReadData <- function(options){
+	  dataset <- .readDataSetToEnd(columns = options[["predictors"]], exclude.na.listwise = options[["predictors"]])
+  if(options[["scaleEqualSD"]] && length(unlist(options[["predictors"]])) > 0)
+    dataset <- .scaleNumericData(dataset)
+	return(dataset)
+}
+
+.mlPredictionModelType <- function(model){
+  type <- NULL
+  if(!is.null(model))
+	type <- switch(class(model),
+						"lda" = "classification",
+						"gbm" = switch(model$distribution$name, "gaussian" = "regression", "multinomial" = "classification"),
+						"randomForest" = model[["type"]],
+						"cv.glmnet" = "regression")
+	return(type)
+}
+
+.mlPredictionReady <- function(model, dataset, options){
+  ready <- TRUE
+  if(!is.null(model)){
+  modelVars <- switch(class(model),
+                      "lda" = decodeColNames(attr(model[["terms"]], "term.labels")),
+                      "gbm" = decodeColNames(attr(model[["Terms"]], "term.labels")),
+                      "randomForest" = decodeColNames(rownames(model[["importance"]])),
+					  "cv.glmnet" = decodeColNames(rownames(model$glmnet.fit$beta)))
+    presentVars <- decodeColNames(colnames(dataset))
+    if(!all(modelVars %in% presentVars)){
+      ready <- FALSE
+    }
+  } else {
+    ready <- FALSE
+  }
+  return(ready)
+}
```

Comment:
This could become:

```suggestion
.mlPredictionGetModelVars <- function(model) {
  useMethod(".mlPredictionGetModelVars", model)
}

.mlPredictionGetModelVars.lda <- function(model) {
  decodeColNames(attr(model[["terms"]], "term.labels"))
}

.mlPredictionGetModelVars.randomForest <- function(model) {
  decodeColNames(rownames(model[["importance"]]))
}

# also define methods for other objects

.mlPredictionReady <- function(model, dataset, options){
  if(!is.null(model)){
    modelVars   <- decodeColNames(mlPredictionGetModelVars(model))
    presentVars <- decodeColNames(colnames(dataset))
    ready <- all(modelVars %in% presentVars))
  } else {
    ready <- FALSE
  }
  return(ready)
}
```